templates/prometheus.yaml.erb

```diff
@@ -7,7 +7,7 @@
 <% full_config = {
   'global'=>global_config,
   'rule_files'=>rule_files,
-  'scrape_configs'=>scrape_configs,
+  'scrape_configs'=>scrape_configs + @collected_scrape_jobs,
```

Comment:
please write a test to verify this line works,

Reply:
do you have any suggestions on where i should start with this? i'm not very familiar with the puppet modules test environment, and even less so when it concerns exported resources...

Reply:
I'm not sure how much you know, so maybe this sounds a bit condescending, but basically you can just:
```
$ bundle install --with test
$ bundle exec rake test
```
And then just go through the stuff in the `spec` folder to get a feel for how things are currently being tested.
As for exported resources, here are some basic examples:
https://github.com/rodjek/rspec-puppet#testing-exported-resources

And yes: I know this is all easier said than done :wink:

Reply:
it's not condescending, I had truly no idea where to start so your feedback was definitely useful. that said:

    bundle install --with test

this is the kind of horrors that make me hesitant of actually getting down that rabbit hole. i modestly imagined this would do a third-party setup of gem things in `/usr/local` but i have actually *zero* idea where all that junk it downloaded was installed or why. it *looks* like it shoved a bunch of things in `/var/lib` but I'm going to pretend all this is perfectly normal and I won't need to worry about arbitrary code being installed as root on my machine. but that's on par for the game these days...

ANYWAY. :)

i tried to do the thing, i really did. here's the patch I came up with:

```diff
modified   spec/defines/daemon_spec.rb
@@ -24,7 +24,10 @@
           group:             'smurf_group',
           env_vars:          { SOMEVAR: 42 },
           bin_dir:           '/usr/local/bin',
-          install_method:    'url'
+          install_method:    'url',
+          export_scrape_job: true,
+          scrape_host:       'localhost',
+          scrape_port:       1234,
         }
       ].each do |parameters|
         context "with parameters #{parameters}" do
@@ -229,6 +232,11 @@
               'enable' => true
             )
           }
+          context 'exported resources' do
+            subject { exported_resources }
+
+            it { is_expected.to contain_file('/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml') }
+          end
         end
       end
     end

```

this fails with:

```
     Failure/Error: it { is_expected.to contain_file('/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml') }
       expected that the catalogue would contain File[/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml]
     # ./spec/defines/daemon_spec.rb:238:in `block (7 levels) in <top (required)>'
```

so the file isn't created? did i do a typo or guessed the name wrong? how do i tell?

... i would love to write test, I really would. but this is just incomprehensible to me. :)

Reply:
The bundle stuff should all go inside the `vendor` folder inside the project. AFAIK it shouldn't install anything system-wide. You mentioned "installed as root" and maybe that's part of the misunderstanding: `bundle` should be called as your normal non-root user. That's what the `$` at the beginning of the commands is meant to indicate. It's a very old convention: `$` indicates the normal user prompt, `#` indicates a root prompt. Just in case you didn't know (again: apologies if this comes over as condescending)

As for the actual test: the `contain_*` part is used to test the existence of some resource in the generated manifest. In this case the exported resource. The file you are testing for is not exported, it is realized by the exported resource, namely `scrape_job`. So here you should be testing if `scrape_job` is contained.
Does this help?

Reply:
I recommend our CONTRIBUTING.md: https://github.com/voxpupuli/puppet-prometheus/blob/master/.github/CONTRIBUTING.md

`bundle install` doesn't install anything as root, unless you prefix it with sudo. But our recommendation is the option `--path .vendor/` which will install all the gems into a subdir within the current module and won't pollute $PATH.

Reply:
i know about `#` and `$`. :) i did run `bundle install` as regular user, but it said it couldn't write to `/var/lib` and asked me for `sudo`... anyways, it turns out it all installed stuff in `/var/lib/gems-2.5.0` which i believe was just empty before (and if it wasn't, it should be). so I did the `--path` trick (thanks for that one, it looks like I hadn't done my homework!)... and I got this error:

```
Downloading listen-3.1.5 revealed dependencies not in the API or the lockfile
(rb-fsevent (>= 0.9.4, ~> 0.9), ruby_dep (~> 1.2)).
Either installing with `--full-index` or running `bundle update listen` should
fix the problem.
```

which i didn't get without `--path`. running with `--full-index` doesn't fix the problem, but `bundle update listen` does, so I plowed ahead.

my tests are still failing. I tried this instead:

```
            it { is_expected.to contain_scrape_job('localhost:1234') }
```

and also:

```
            it { is_expected.to contain_prometheus_scrape_job('localhost:1234') }
```

they both fail with an error similar to the original:

```
     Failure/Error: it { is_expected.to contain_prometheus_scrape_job('localhost:1234') }
       expected that the catalogue would contain Prometheus_scrape_job[localhost:1234]
     # ./spec/defines/daemon_spec.rb:238:in `block (7 levels) in <top (required)>'
```

is there a way to list the resources that are actually exported? where are those tests running anyways? how can I debug this?

thanks for the hand holding! :)

Reply:
For `contain_*` examples, see [here](https://rspec-puppet.com/documentation/defined_types/#test-a-resource). In this specific case you should probably be using `contain_prometheus__scrape_job` (note the "dunder"/double underscore to denote "::").

As for @bastelfreak 's original comment: yes, you'd have to also test the contents of the generated file to be sure the concatenation makes sense. This should be done in a separate test case, though.

Reply:
alright, i pushed what i think is a working test case for `scrape_job.pp`, let me know if that looks alright to you:

```ruby
          context 'exported resources' do
            subject { exported_resources }
            it {
              is_expected.to contain_prometheus__scrape_job('localhost:1234')
            }
          end
```

also pushed here...

I did try to test the original line commented here, but I couldn't figure out how to create the scrape_job resources (the daemon, really) from the other file. I tried to add this to `classes/prometheus_rspec.rb`:

```ruby
      context 'exported resources' do
        [
          {
            collect_scrape_jobs: [
              { job_name: 'smurf_exporter' },
            ]
          }
        ].each do |parameters|
          context "with prometheus version #{parameters[:version]}" do
            let(:params) do
              parameters
            end
            # copied over from defines/daemon_spec.rb
            let(:pre_condition) { 'prometheus::daemon { "smurf_exporter": export_scrape_job => true, scrape_host => "localhost", scrape_port => 1234, real_download_url => "https://github.com/prometheus/smurf_exporter/releases/v1.2.3/smurf_exporter-1.2.3.any.tar.gz"' }
            subject { exported_resources }
            it {
              is_expected.to contain_prometheus__scrape_job('localhost:1234')
            }
          end
        end
      end
```

but that crashes with:

```
  6) prometheus on ubuntu-18.04-x86_64 exported resources with prometheus version  should contain Prometheus::Scrape_job[localhost:1234]
     Failure/Error: is_expected.to contain_prometheus__scrape_job('localhost:1234')
     
     Puppet::ParseErrorWithIssue:
       Could not parse for environment rp_env: Syntax error at 'class' (line: 3, column: 1) on node curie.anarc.at
```

... since i'm just cargo culting this, i have no idea why it fails again. ;) 

i guess that rspec is creating a puppet manifest and applying it with that stuff - is there a place where I can actually see what manifest it's creating to debug what i'm doing wrong?

thanks again for all the hand holding! we're getting closer!